### PR TITLE
Reduce call get available sites for TYPO3 12.4

### PR DESCRIPTION
**Pull request by sfroemkenjw**
What this pr does?

Use PHP generator with yield to process just the available sites needed
How to test

With over 300 root pages it needs over 30 seconds to build up all available sites. With help of a PHP generator we can stop processing all available sites if just the first site is requested. It's also helpful to just check, if there are available sites.

This should speed up performance a lot.

### CODE REVIEW

**Comment by sfroemkenjw** [comment]
Related to PR: #4145

**Comment by sfroemkenjw** [comment]
Hi @dkd-kaehm 

I don't understand. After your approval you will get 3 options to merge my PR: Merge, Rebase and Squash. In TYPO3 docs team we always work this way. I have no clue how to squash commits in my side.

Stefan

**Comment by dkd-kaehm** [comment]
OK I'll do the squash.
I have a question on one change.

File: Tests/Unit/Domain/Site/SiteRepositoryTest.php [hunk 1]
```diff
@@ -89,7 +89,6 @@ public function canGetFirstAvailableSite(): void
         $this->assertThatSitesAreCreatedWithPageIds([333], [
             0 => ['language' => 0],
         ]);
-        $this->assertCacheIsWritten();
```
Comment (dkd-kaehm) [inline Tests/Unit/Domain/Site/SiteRepositoryTest.php:92]: Why removing that assertion?

Comment (sfroemkenjw) [reply Tests/Unit/Domain/Site/SiteRepositoryTest.php:92]: Caching of available site only happens while calling `getAvailableSites`. But as `hasAvailableSites` works with the generator directly "typo3ManagedSolrSites", no caching happens. That's why I have removed that check.

**Comment by dkd-kaehm** [comment]
Squashed and rebased to release-12.0.x on 2024.12.12 state.

**Comment by dkd-kaehm** [comment]
Replaced by #4260

**Pull request closed without merging**